### PR TITLE
Added diagnostics to sqlclient

### PR DESCRIPTION
**Pull request by tstringer**
This pull request includes changes to SqlClient in order to provide a mechanism for facilitating diagnostics collection in the library for all `Execute*()` operations (sync and async).

Tests have been created in order to test the functionality as well.

This functionality is proposed for merging as a facilitator of client-side diagnostics and logging to collect relevant data retrieval information for monitoring, baselining, and troubleshooting.

### CODE REVIEW

**Comment by saurabh500** [comment]
@tstringer all tests are failing. Can you take a look?

**Comment by tstringer** [comment]
@saurabh500 without further investigation, I'm assuming that the SqlClient tests are failing largely due to the fact that they require a connection.  I took the connection string from the ExceptionTests (https://github.com/tstringer/corefx/blob/sqlclient-diagnostics/src/System.Data.SqlClient/tests/FunctionalTests/ExceptionTest.cs#L14) and used it in my tests: https://github.com/tstringer/corefx/blob/sqlclient-diagnostics/src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs#L13 for reference.  Do you know if that is a connect-able source?  Any recommendation on these connection strings to actually connect to a valid data source?

**Comment by tstringer** [comment]
@saurabh500 yep, looks like that's the cause: `System.Data.SqlClient.SqlException : A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections.`

Let me know what connection string I can use in my tests for the CI builds! [Tests' connection string can be found here, it appears this is invalid](https://github.com/tstringer/corefx/blob/sqlclient-diagnostics/src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs#L13)

**Comment by nbilling** [comment]
The exception tests don't expect the connection string to work since they all get into one error state or another before the connection string actually gets used.

We already discussed the unit test thing and it would require a SqlConnection mock that doesn't currently exist in corefx. Our decision was to go without any unit tests and to either change those tests to manual ones or omit them. Possibly I forgot to pass this on to you @tstringer ? My bad if so.

**Comment by avanderhoorn** [comment]
@saurabh500 if the decision is to omit these... do we just want them comments out/ignored for now so that they are there when you do get a mock connection eventually?

**Comment by tstringer** [comment]
@nbilling @avanderhoorn thanks for the input, that sounds like the way to go here.  @saurabh500 how would you like me to proceed here?  Would you like for me to make the test attributes `[Fact(Skip = "Skipping due to data source dependency")]`?  Let me know and I'll make those changes asap! :smile:

**Comment by saurabh500** [comment]
The connection string for exception test is definitely not expected to work :) 

For SqlClient we don't have any tests connecting to the server in Github. We do run tests internally for SqlClient using on premise and Azure Sql servers to make sure things are not broken. It is not a good solution and needs work.

@tstringer You have a few options.
1. You can disable the tests. Check in a Readme on how to execute these tests locally so that we can verify the functionality. We do something similar for our ManualTests folder. 
2. You can use localdb to run the tests on Windows to get some coverage. EF team uses LocalDB for some of their tests. That would work as well.

Option 2 would be preferred option. You could start with Option 1, open an issue to migrate to Option 2 and send another PR with Option 2 incorporated into the tests. 

Long term, we will consider open sourcing TDS server and setting up Azure Sql / Sql in a VM infra for testing.

File: src/System.Data.SqlClient/tests/FunctionalTests/project.json [hunk 1]
```diff
@@ -4,6 +4,11 @@
     "Microsoft.NETCore.Platforms": "1.0.1-rc3-24109-00",
     "System.Data.Common": "4.0.1-rc3-24109-00",
     "System.Text.RegularExpressions": "4.1.0-rc3-24109-00",
+    "System.Collections": "4.0.11-rc3-24109-00",
+    "System.Collections.Concurrent": "4.0.12-rc3-24109-00",
+    "System.Collections.NonGeneric": "4.0.1-rc3-24109-00",
```
Comment (saurabh500) [inline src/System.Data.SqlClient/tests/FunctionalTests/project.json]: Is this dependency necessary? We are trying to move away from `System.Collection.NonGeneric`

Comment (tstringer) [inline src/System.Data.SqlClient/tests/FunctionalTests/project.json]: @saurabh500 That dependency is unnecessary, and I have removed it (and tested post-removal) in commit https://github.com/dotnet/corefx/pull/8420/commits/23de00d65e4e80cf7e28be84a797be906912c692

**Comment by tstringer** [comment]
@saurabh500 I have changed the diagnostics tests to be run manually, and this change is reflected in commit https://github.com/dotnet/corefx/pull/8420/commits/5de34d82af83389becf6db258245a49fb4b41d01

**Comment by nbilling** [comment]
Should the test be moved from functional to manual directory?

**Comment by tstringer** [comment]
@nbilling I had a quick phone conversation with @saurabh500 and the intent here is to get through this PR like this, and then I'll create an issue and subsequent PR that makes these tests hit localdb so they are functional at least on Windows.  So this is a temporary fix for these tests at the moment that will be resolved shortly.

**Comment by nbilling** [comment]
Fair enough.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlClientDiagnosticListenerExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    /// Extension methods on the DiagnosticListener class to log SqlCommand data
+    /// </summary>
+    internal static class SqlClientDiagnosticListenerExtensions
+    {
+        public static readonly string DiagnosticListenerName = "SqlClientDiagnosticListener";
+
+        private static readonly string SqlClientPrefix = "System.Data.SqlClient.";
+        public static readonly string SqlBeforeExecuteCommand = SqlClientPrefix + nameof(WriteCommandBefore);
+        public static readonly string SqlAfterExecuteCommand = SqlClientPrefix + nameof(WriteCommandAfter);
+        public static readonly string SqlErrorExecuteCommand = SqlClientPrefix + nameof(WriteCommandError);
```
Comment (justinvp) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlClientDiagnosticListenerExtensions.cs]: Is there a reason these need to be `static readonly` fields, or could they be constants? e.g.:

```csharp
public const string DiagnosticListenerName = "SqlClientDiagnosticListener";

private const string SqlClientPrefix = "System.Data.SqlClient.";
public const string SqlBeforeExecuteCommand = SqlClientPrefix + nameof(WriteCommandBefore);
public const string SqlAfterExecuteCommand = SqlClientPrefix + nameof(WriteCommandAfter);
public const string SqlErrorExecuteCommand = SqlClientPrefix + nameof(WriteCommandError);
```

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs [hunk 3]
```diff
@@ -290,7 +293,8 @@ internal SqlStatistics Statistics
             {
                 if (null != _activeConnection)
                 {
-                    if (_activeConnection.StatisticsEnabled)
+                    if (_activeConnection.StatisticsEnabled || 
+                        _diagnosticListener.IsEnabled(SqlClientDiagnosticListenerExtensions.SqlAfterExecuteCommand))
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: What "enables" the diagnosticListener? Also how can tihs be disabled?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs [hunk 4]
```diff
@@ -523,6 +526,19 @@ override public void Open()
         {
             PrepareStatisticsForNewConnection();
 
+            if (StatisticsEnabled ||
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs]: The `PrepareStatisticsForNewConnection` method initializes the _statistics as well. Why is `_statistics` initialization needed again?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs [hunk 5]
```diff
@@ -748,6 +764,19 @@ private void CancelOpenAndWait()
         public override Task OpenAsync(CancellationToken cancellationToken)
         {
             PrepareStatisticsForNewConnection();
+            
+            if (StatisticsEnabled ||
+                _diagnosticListener.IsEnabled(SqlClientDiagnosticListenerExtensions.SqlAfterExecuteCommand))
+            {
+                if (null == _statistics)
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs]: The PrepareStatisticsForNewConnection method (called above) initializes the _statistics as well. Why is _statistics initialization needed again?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs (hunk 3 above)
Comment (nbilling) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: A listener is _enabled_ when an observer is registered and self-reports as listening to the diagnosticListener. So in practice it will be disabled by default and then enabled once an interested observer shows up.

Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: @nbilling How is the observer registered? Any samples on the usage of this capability? 
What code / class should I look at, to see what the observer is going to look like?

Comment (nbilling) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: In the tests there is a toy observer called `FakeDiagnosticListenerObserver `. In the tests it gets hooked up to the SqlClient DiagnosticListener on line `DiagnosticListener.AllListeners.Subscribe(diagnosticListenerObserver);` (where `diagnosticListenerObserver` is a `FakeDiagnosticListenerObserver`).

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs [hunk 6]
```diff
@@ -40,6 +40,9 @@ public sealed partial class SqlConnection : DbConnection
         internal bool _supressStateChangeForReconnection;
         private int _reconnectCount;
 
+        // diagnostics listener
+        private readonly static DiagnosticListener _diagnosticListener = new DiagnosticListener(SqlClientDiagnosticListenerExtensions.DiagnosticListenerName);
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs]: Nit: `s_diagnosticListener` ?

File: src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+        [Fact]
+        private void ExecuteXmlReaderErrorTest()
+        {
+            CollectStatisticsDiagnostics(ExecuteXmlReaderError);
+        }
+
+        [Fact]
+        private void ExecuteScalarAsyncTest()
+        {
+            //await ExecuteScalarAsync();
```
Comment (nbilling) [inline src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs]: Delete?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlClientDiagnosticListenerExtensions.cs (hunk 2 above)
Comment (tstringer) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlClientDiagnosticListenerExtensions.cs]: @justinvp I have made these changes and they are reflected in commit https://github.com/dotnet/corefx/pull/8420/commits/319736ba1954415602e7aecf25ba52daee275f36 Thanks! :smile:

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs (hunk 3 above)
Comment (tstringer) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: Thanks, @nbilling, for answering these questions!  @saurabh500 let me know if you have any more questions about this.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs (hunk 4 above)
Comment (tstringer) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs]: @saurabh500 I have refactored this logic into `PrepareStatisticsForNewConnect()` to eliminate the redundancy.  These changes can be found in commit https://github.com/dotnet/corefx/pull/8420/commits/319736ba1954415602e7aecf25ba52daee275f36 Thanks! :smile:

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs (hunk 5 above)
Comment (tstringer) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs]: @saurabh500 I have refactored this logic into `PrepareStatisticsForNewConnect()` to eliminate the redundancy.  These changes can be found in commit https://github.com/dotnet/corefx/pull/8420/commits/319736ba1954415602e7aecf25ba52daee275f36 Thanks! :smile:

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs (hunk 6 above)
Comment (tstringer) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs]: @saurabh500 I have changed the name to `s_diagnosticListener` and this change can be found in commit https://github.com/dotnet/corefx/pull/8420/commits/319736ba1954415602e7aecf25ba52daee275f36 Thanks! :smile:

File: src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs (hunk 7 above)
Comment (tstringer) [inline src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs]: @nbilling I've removed this unnecessary comment and this change can be found in commit https://github.com/dotnet/corefx/pull/8420/commits/319736ba1954415602e7aecf25ba52daee275f36 Thanks! :smile:

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs [hunk 8]
```diff
@@ -738,19 +742,35 @@ override public object ExecuteScalar()
...
             {
                 SqlStatistics.StopTimer(statistics);
+
+                if (e != null)
+                {
+                    _diagnosticListener.WriteCommandError(operationId, this, nameof(ExecuteScalar), e);
+                }
+                else
+                {
+                    _diagnosticListener.WriteCommandAfter(operationId, this, nameof(ExecuteScalar));
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: Should WriteCommandAfter take an Exception argument instead?  Then these calls condense down to:
```C#
_diagnosticListener.WriteCommandAfter(operationId, this, e, nameof(ExecuteScalar));
```
rather than having the branching at every call site.  Further, you could use a [CallerMemberName] argument to avoid needing to specify the function name explicitly, so it would just become:
```C#
_diagnosticListener.WriteCommandAfter(operationId, this, e);
```
and for call sites where there is no Exception, null can be passed, and the Exception argument could even be made optional with a default value of null.

File: src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs [hunk 9]
```diff
@@ -0,0 +1,507 @@
...
+
+                        string operation = GetPropertyValueFromType<string>(kvp.Value, "Operation");
+                        Assert.False(string.IsNullOrWhiteSpace(operation));
+
+                        statsLogged = true;
+                    }
+                });
+
+            diagnosticListenerObserver.Enable();
+            DiagnosticListener.AllListeners.Subscribe(diagnosticListenerObserver);
```
Comment (stephentoub) [inline src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs]: We should dispose of the IDisposable returned by Subscribe before exiting the test.

File: src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs [hunk 10]
```diff
@@ -0,0 +1,507 @@
...
+
+                    string operation = GetPropertyValueFromType<string>(kvp.Value, "Operation");
+                    Assert.False(string.IsNullOrWhiteSpace(operation));
+
+                    statsLogged = true;
+                }
+            });
+
+            diagnosticListenerObserver.Enable();
+            DiagnosticListener.AllListeners.Subscribe(diagnosticListenerObserver);
```
Comment (stephentoub) [inline src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs]: Same.

File: src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs [hunk 11]
```diff
@@ -0,0 +1,507 @@
...
+
+            await sqlOperation();
+
+            Assert.True(statsLogged);
+
+            diagnosticListenerObserver.Disable();
+        }
+
+        private void ExecuteScalar()
```
Comment (stephentoub) [inline src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs]: Why have this separate from the [Fact]?  Rather than:
```C#
[Fact]  
public void ExecuteScalarTest()  
{
    CollectStatisticsDiagnostics(ExecuteScalar);
}
... // lots of code
private void ExecuteScalar()
{
    using (SqlConnection conn = new SqlConnection(_connectionString)
    using (SqlCommand cmd = new SqlCommand())
    {
        ... // etc.
    }
}
```
you could just have:
```C#
[Fact]  
private void ExecuteScalarTest()  
{
    CollectStatisticsDiagnostics(() =>
    {
        using (SqlConnection conn = new SqlConnection(_connectionString)
        using (SqlCommand cmd = new SqlCommand())
        {
            ... // etc.
        }
    });
}
```

File: src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs [hunk 12]
```diff
@@ -0,0 +1,507 @@
...
+        private string _connectionString = "server=tcp:server,1432;database=test;uid=admin;pwd=SQLDB;connect timeout=60;";
+        private readonly ITestOutputHelper _output;
+
+        public DiagnosticTest(ITestOutputHelper output)
+        {
+            _output = output;
+        }
+
+        [Fact]
+        private void ExecuteScalarTest()
```
Comment (stephentoub) [inline src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs]: Do these tests execute when they're private?  I was under the impression they needed to be public for xunit to pick them up, but maybe I'm wrong or maybe my knowledge on that is out-of-date.

Comment (nbilling) [inline src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs]: They were running for me when I did a full build from the root of corefx. I didn't even notice they are private. Weird.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs (hunk 8 above)
Comment (tstringer) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: @stephentoub In my personal opinion, the clear delineation between `*After` and `*Error` in the API is a true portrayal of not only the different logged events, but also give us two different code paths to minimize spaghetti logic.  The data points collected for these two events only have a couple of things in common, whereas the `*Error` event collects [`Exception`](https://github.com/tstringer/corefx/blob/sqlclient-diagnostics/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlClientDiagnosticListenerExtensions.cs#L67) and the `*After` event collects [`Statistics`](https://github.com/tstringer/corefx/blob/sqlclient-diagnostics/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlClientDiagnosticListenerExtensions.cs#L50).  Whereas I can understand your thoughts to merge this into a single method, I think the entry point is really the only common part of these two events.

In summary, going forward I think having both methods gives a clear indication of future calls.  With that being said, I'd be curious to get @avanderhoorn's thoughts on this.

File: src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs (hunk 11 above)
Comment (tstringer) [inline src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs]: @stephentoub that's a good suggestion, I can refactor these tests with your recommendation.  Thanks! :smile:

File: src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs (hunk 12 above)
Comment (tstringer) [inline src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs]: @stephentoub yes private methods are picked up and tested by xunit.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs (hunk 8 above)
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: Whatever you collectively decide is fine.  From my perspective, though, it's much cleaner to have a Before and After, where you always invoke both of those, with the After will be invoked regardless of whether completion is successful.  It can then internally branch to spit out the correct underlying events for the situation.  This maps more cleanly to how I think about it and also helps to keep the call sites clean... but again, that's just my opinion.

File: src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs (hunk 12 above)
Comment (stephentoub) [inline src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs]: Ok, thanks.  FWIW, for consistency with the rest of the repo, I'd suggest making the [Fact] methods public.  Otherwise I believe these will be the only ones in the whole repo that are private ;)

Comment (tstringer) [inline src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs]: @stephentoub Sounds good, I'll make those changes to ensure consistency! :smile:

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs (hunk 8 above)
Comment (avanderhoorn) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: @stephentoub I like having the two separate, but you feel feel strongly about it, then we can handle the difference within the `WriteCommandAfter` and publish our two different events based on exception being null or not. I can see that this change would simplify the code at this site and in the long run I can see the benefit here for that.

Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: As the owner of the lib, @saurabh500 and co should decide.  I'm just providing my $.02 ;)

Comment (tstringer) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: @stephentoub :smile: sounds good and much appreciated!  I'll take the route that @saurabh500 would like to see here.

Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: @stephentoub  Thanks for the review :smile:  
I think it makes more sense to have two different methods in this case. 

@tstringer Can you use [CallerMemberName]  as suggested by @stephentoub . It keeps the code cleaner.

**Comment by saurabh500** [comment]
@tstringer I went through the Async changes and it looks good to me. :+1: 
The only open item right now is the [CallerMemberName] comment. You are good to merge after that. 
Do you have permissions or do you need us to merge?

**Comment by tstringer** [comment]
@saurabh500 Great!  I have made the changes to incorporate `[CallerMemberName]` and now all of the requested changes should be reflected in the commit.  Of note is that there were merge conflicts this morning, so I did a pull from upstream master and rebased this branch.  I also squashed the commits into a single commit.

To answer your question directly, unfortunately no I don't have write access to this repository, so I would need you to merge this pull request.

Thanks! :smile:

**Comment by avanderhoorn** [comment]
Thanks so much to everyone involved for making this happen!!! This was a gigantic effort and will make a big difference. Thanks again to @tstringer making the code happen, @saurabh500 & @YoungGah for being open to getting this code in and taking the time to review, @nbilling for your second pair of eyes, and @justinvp @stephentoub for your input/reviews!

**Comment by tstringer** [comment]
**Huge** thanks to the entire team, especially for the code reviews!!  It was a pleasure working with everybody, and I look forward to further contributions and interactions with everyone involved here!
